In [1]:
import pandas as pd

In [2]:
import graphlab as gl
from graphlab import SFrame
import graphlab as gl

In [3]:
from graphlab.toolkits.recommender import factorization_recommender

In [4]:
df = pd.read_pickle('nutrition.pkl')

First Model trained is simply checking to root mean squared error on the SFrame based solely on user, generalizing item by brand, and included what they rated it. Base RMSE found to be 0.95111

In [33]:
m01 = factorization_recommender.create(sf_train[['userId', 'brandId', 'rating']], user_id = 'userId', item_id='brandId', target='rating')


Recsys training: model = factorization_recommender

Preparing data set.

Data has 46199 observations with 31023 users and 323 items.

Data prepared in: 0.178463s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | sgd      |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 10000 / 46199 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 25                | Not Viable                               |

| 1       | 6.25              | Not Viable                               |

| 2       | 1.5625            | Not Viable                               |

| 3       | 0.390625          | No Decrease (4.69996 >= 4.57544)         |

| 4       | 0.0976562         | 0.956562                                 |

| 5       | 0.0488281         | 2.60204                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.0976562         | 0.956562                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 109us        | 4.52543           | 2.12731               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 90.681ms     | 4.65336           | 2.15714               | 0.0976562   |

| 2       | 166.611ms    | 2.89494           | 1.70141               | 0.0580668   |

| 3       | 239.034ms    | 1.45511           | 1.20622               | 0.042841    |

| 4       | 293.816ms    | 0.76834           | 0.876464              | 0.0290334   |

| 5       | 344.668ms    | 0.537441          | 0.733                 | 0.0245592   |

| 6       | 510.47ms     | 0.430721          | 0.656174              | 0.0214205   |

| 11      | 996.324ms    | 0.272658          | 0.522008              | 0.0135956   |

| 25      | 2.16s        | 0.210268          | 0.458362              | 0.00734493  |

| 50      | 3.65s        | 0.189209          | 0.43478               | 0.00436732  |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.170648

Final training RMSE: 0.412883

In [24]:
eval = m01.evaluate_rmse(sf_test[['userId', 'brandId', 'rating']], target= 'rating')
eval

{'rmse_by_item': Columns:
 	brandId	str
 	count	int
 	rmse	float
 
 Rows: 250
 
 Data:
 +-------------------------+-------+----------------+
 |         brandId         | count |      rmse      |
 +-------------------------+-------+----------------+
 |   BRAND_DOCTOR_S_BEST   |   14  | 3.07602114693  |
 |  BRAND_NORDIC_NATURALS  |   7   | 1.52659685652  |
 |  BRAND_ALLMAX_NUTRITION |   63  | 3.05039978569  |
 |        cat1340006       |   6   | 2.25337669137  |
 |         BRAND_GU        |   6   |  1.4955570503  |
 |        cat320010        |  384  | 2.22269357873  |
 | BRAND_ENZYMATIC_THERAPY |   5   | 0.381342089108 |
 |        cat510018        |   6   | 2.27351097181  |
 |        cat1030006       |   17  | 3.54611578851  |
 |        cat1960001       |   1   | 0.289701709895 |
 +-------------------------+-------+----------------+
 [250 rows x 3 columns]
 Note: Only the head of the SFrame is printed.
 You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.,
 '

In [56]:
m09 = factorization_recommender.create(sf_train[['userId', 'productId', 'rating']], user_id = 'userId', item_id='productId', target='rating')

Recsys training: model = factorization_recommender

Preparing data set.

Data has 46169 observations with 31073 users and 1992 items.

Data prepared in: 0.175567s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | sgd      |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 10000 / 46169 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 25                | Not Viable                               |

| 1       | 6.25              | Not Viable                               |

| 2       | 1.5625            | No Decrease (64.0795 >= 4.7443)          |

| 3       | 0.390625          | 0.438626                                 |

| 4       | 0.195312          | 0.604749                                 |

| 5       | 0.0976562         | 1.39051                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.390625          | 0.438626                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 124us        | 4.63679           | 2.15332               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 49.7ms       | 8.32001           | 2.88419               | 0.390625    |

| 2       | 122.582ms    | DIVERGED          | DIVERGED              | 0.232267    |

| RESET   | 148.097ms    | 4.63686           | 2.15334               |             |

| 1       | 196.405ms    | 4.58592           | 2.14146               | 0.116134    |

| 2       | 262.056ms    | 3.14309           | 1.77286               | 0.0690534   |

| 3       | 305.045ms    | 2.23631           | 1.49541               | 0.0509468   |

| 4       | 367.874ms    | 1.50595           | 1.22715               | 0.0410594   |

| 5       | 445.411ms    | 0.971608          | 0.985669              | 0.034732    |

| 10      | 734.674ms    | 0.214987          | 0.463557              | 0.0206518   |

| 23      | 1.31s        | 0.0425562         | 0.205991              | 0.0110576   |

| 48      | 2.81s        | 0.0126099         | 0.111706              | 0.00636835  |

| 50      | 2.91s        | 0.0118161         | 0.108093              | 0.00617632  |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.011111

Final training RMSE: 0.104781

In [57]:
eval = m09.evaluate_rmse(sf_test[['userId', 'productId', 'rating']], target= 'rating')
eval

{'rmse_by_item': Columns:
 	productId	str
 	count	int
 	rmse	float
 
 Rows: 1519
 
 Data:
 +-------------+-------+----------------+
 |  productId  | count |      rmse      |
 +-------------+-------+----------------+
 | prod1250008 |   5   | 3.76805491352  |
 |  prod140038 |   8   | 0.798663481768 |
 |    25154    |   2   | 1.25704123067  |
 |    26493    |   3   | 2.83606805901  |
 |    26994    |   86  | 1.35366066188  |
 |    28496    |   3   | 1.86522803519  |
 |    26713    |   11  | 0.890392337044 |
 | prod3820059 |   86  | 1.15838625538  |
 |  prod800014 |   3   | 1.94012446024  |
 | prod1180014 |   1   | 4.23313754661  |
 +-------------+-------+----------------+
 [1519 rows x 3 columns]
 Note: Only the head of the SFrame is printed.
 You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.,
 'rmse_by_user': Columns:
 	userId	str
 	count	int
 	rmse	float
 
 Rows: 12792
 
 Data:
 +----------+-------+----------------+
 |  userId  | count |      rmse      |


In [10]:
df[df.weight > 0].weight.mean(), df[df.height > 0].height.mean(), df[df.bodyfat > 0].bodyfat.mean()

(181.77160511611612, 69.52690519331404, 15.524334234187396)

In [65]:
user_info = sf_train[['userId', 'height', 'weight', 'bodyfat']]
item_info = sf_train[['brandId', 'productId']]
user_info = user_info.fillna('weight',181.7716)
user_info = user_info.fillna('height',69.5269)
user_info = user_info.fillna('bodyfat',15.5243)
m07 = factorization_recommender.create(sf_train[['userId', 'productId', 'rating']], user_id = 'userId', item_id='productId', target='rating',\
                                    #user_data=user_info,\
                                    item_data=item_info)

Recsys training: model = factorization_recommender

Preparing data set.

Data has 46170 observations with 31071 users and 2008 items.

Data prepared in: 0.148373s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 10000 / 46170 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 16.6667           | Not Viable                               |

| 1       | 4.16667           | No Decrease (21.5597 >= 4.58588)         |

| 2       | 1.04167           | 0.53367                                  |

| 3       | 0.520833          | 0.149768                                 |

| 4       | 0.260417          | 0.216257                                 |

| 5       | 0.130208          | 0.619232                                 |

+---------+-------------------+------------------------------------------+

| Final   | 0.520833          | 0.149768                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 119us        | 4.54844           | 2.13271               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 41.548ms     | 4.80553           | 2.19195               | 0.520833    |

| 2       | 113.945ms    | 2.61072           | 1.61552               | 0.520833    |

| 3       | 152.463ms    | 1.06709           | 1.03262               | 0.520833    |

| 4       | 212.41ms     | 0.456048          | 0.674728              | 0.520833    |

| 5       | 264.125ms    | 0.212475          | 0.460092              | 0.520833    |

| 6       | 326.643ms    | 0.111294          | 0.332419              | 0.520833    |

| 11      | 726.339ms    | 0.0229296         | 0.148779              | 0.520833    |

| 25      | 1.48s        | 0.0087273         | 0.0890617             | 0.520833    |

| 50      | 3.21s        | 0.010719          | 0.0996324             | 0.520833    |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.0143545

Final training RMSE: 0.116456

In [66]:
eval = m07.evaluate_rmse(sf_test[['userId', 'productId', 'rating', 'brandId']], target= 'rating')
eval

{'rmse_by_item': Columns:
 	productId	str
 	count	int
 	rmse	float
 
 Rows: 1519
 
 Data:
 +-------------+-------+----------------+
 |  productId  | count |      rmse      |
 +-------------+-------+----------------+
 | prod1250008 |   6   |  4.1462266865  |
 |  prod140038 |   9   | 1.24685206005  |
 |    25154    |   1   | 1.98425638797  |
 |    26493    |   1   | 1.79682567595  |
 |    26994    |  100  | 1.47388070783  |
 |    28496    |   3   | 2.66383027909  |
 | prod1120036 |   1   | 2.15727895377  |
 |    26713    |   13  | 0.996572648736 |
 | prod3820059 |   68  | 1.41687581438  |
 |  prod800014 |   1   |  1.1567268881  |
 +-------------+-------+----------------+
 [1519 rows x 3 columns]
 Note: Only the head of the SFrame is printed.
 You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.,
 'rmse_by_user': Columns:
 	userId	str
 	count	int
 	rmse	float
 
 Rows: 12869
 
 Data:
 +-----------+-------+----------------+
 |   userId  | count |      rmse      

In [101]:
sf = SFrame(df)
sf_train, sf_test = sf.random_split(.75)

Item     Item Data  User Data         Train    Test
Brand    None       None              .412    2.302
Product  None       None              .105    2.113
Product  Brand      None              .116    2.230
Product  None       Weight            .132    2.284
Product  None       Height           1.087    2.390
Product  None       Bodyfat           .450    2.373
Product  None       Weight, Bf        .592    2.855
Product  None       Height, Bf        .856    2.681
Product  None       Height, Weight   2.087    2.132
Product  None       Ht, Wt, Bf       1.870    2.256


In [103]:
user_info = sf_train[['userId', 'height', 'weight', 'bodyfat']]
item_info = sf_train[['productId']]
user_info = user_info.fillna('weight',181.7716)
user_info = user_info.fillna('height',69.5269)
user_info = user_info.fillna('bodyfat',15.5243)
lb = factorization_recommender.create(sf_train[['userId', 'productId', 'rating']], user_id = 'userId', item_id='productId', target='rating',\
                                     user_data=user_info)#,\
#                                     item_data=item_info)

Recsys training: model = factorization_recommender

Preparing data set.

Data has 46367 observations with 31161 users and 1973 items.

Data prepared in: 0.685639s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 10000 / 46367 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 10                | Not Viable                               |

| 1       | 2.5               | Not Viable                               |

| 2       | 0.625             | Not Viable                               |

| 3       | 0.15625           | No Decrease (5.18042 >= 4.67807)         |

| 4       | 0.0390625         | 3.57296                                  |

| 5       | 0.0195312         | 4.20784                                  |

| 6       | 0.00976562        | 4.43192                                  |

| 7       | 0.00488281        | 4.55595                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.0390625         | 3.57296                                  |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 135us        | 4.56799           | 2.13729               |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 86.448ms     | 4.4558            | 2.11088               | 0.0390625   |

| 2       | 192.545ms    | 4.24682           | 2.06078               | 0.0390625   |

| 3       | 257.925ms    | 4.08696           | 2.02162               | 0.0390625   |

| 4       | 330.478ms    | 3.94442           | 1.98605               | 0.0390625   |

| 5       | 406.23ms     | 3.79457           | 1.94796               | 0.0390625   |

| 6       | 491.468ms    | 3.63034           | 1.90534               | 0.0390625   |

| 11      | 831.107ms    | 2.93838           | 1.71416               | 0.0390625   |

| 25      | 2.19s        | 1.77172           | 1.33103               | 0.0390625   |

| 50      | 4.24s        | 1.28716           | 1.13448               | 0.0390625   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 1.24618

Final training RMSE: 1.11627

In [105]:
eval = lb.evaluate_rmse(sf_test[['userId', 'productId','height', 'weight', 'bodyfat', 'rating']], target= 'rating')
eval

[ERROR] graphlab.toolkits._main: Toolkit error: Provided data missing required columns: productId.


ToolkitError: Provided data missing required columns: productId.

Next model trained incorporated user specific attributes as well as item specific data, effectively lowering the RMSE to 0.261932

In [332]:
df = pd.read_pickle('nutrition.pkl')

In [337]:
sf = gl.SFrame(df[['userId', 'productId', 'rating']])
user_info = gl.SFrame(df[['username', 'userId', 'height', 'weight', 'bodyfat']])
item_info = gl.SFrame(df[['brandName', 'name', 'brandId', 'productId']])

In [343]:
user_info = user_info.fillna('weight',0)
user_info = user_info.fillna('height',0)
user_info = user_info.fillna('bodyfat',0)



In [344]:
m = factorization_recommender.create(sf, user_id='userId', item_id='productId', target='rating',\
                                     user_data=user_info, item_data=item_info)

Recsys training: model = factorization_recommender

Preparing data set.

Data has 64312 observations with 39465 users and 2131 items.

Data prepared in: 0.577575s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 10000 / 64312 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 5.55556           | Not Viable                               |

| 1       | 1.38889           | Not Viable                               |

| 2       | 0.347222          | Not Viable                               |

| 3       | 0.0868056         | 1.56668                                  |

| 4       | 0.0434028         | 2.05017                                  |

| 5       | 0.0217014         | 3.23899                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.0868056         | 1.56668                                  |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 194us        | 4.54328           | 2.1315                |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 240.92ms     | 4.37119           | 2.09072               | 0.0868056   |

| 2       | 394.357ms    | 3.21557           | 1.79318               | 0.0868056   |

| 3       | 570.799ms    | 2.34373           | 1.53089               | 0.0868056   |

| 4       | 710.548ms    | 1.71776           | 1.31059               | 0.0868056   |

| 5       | 808.914ms    | 1.29278           | 1.13695               | 0.0868056   |

| 6       | 887.623ms    | 1.0056            | 1.00273               | 0.0868056   |

| 10      | 1.19s        | 0.426159          | 0.652679              | 0.0868056   |

| 11      | 1.27s        | 0.359798          | 0.599687              | 0.0868056   |

| 20      | 2.35s        | 0.144336          | 0.37966               | 0.0868056   |

| 30      | 3.88s        | 0.0927258         | 0.304176              | 0.0868056   |

| 40      | 5.19s        | 0.0627754         | 0.250139              | 0.0868056   |

| 50      | 6.03s        | 0.0516047         | 0.22671               | 0.0868056   |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.0488329

Final training RMSE: 0.220512

In [177]:
m.show()

Canvas is accessible via web browser at the URL: http://localhost:60407/index.html
Opening Canvas in default web browser.


In [51]:
import graphlab as gl
from sklearn.decomposition import NMF, LatentDirichletAllocation
import pandas as pd
from nltk.corpus import stopwords
from nltk.stem.snowball import SnowballStemmer
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from string import punctuation

In [52]:
def topics(n_topics):
    corpus = []
    stemmer = SnowballStemmer('english')
    for prodId in df.productId.unique():
        docs = []
        for doc in df[df.productId == prodId].text:
            if doc:
                doc = ' '.join(stemmer.stem(word.strip(punctuation).lower()) \
                                                for word in doc.split())
                docs.append(doc)
        corpus.append(' '.join(docs))
    tfidf_vectorizer = TfidfVectorizer(stop_words='english')
    tfidf = tfidf_vectorizer.fit_transform(corpus)
    nmf = NMF(n_components = n_topics).fit(tfidf)
    n_topic_words = n_topics
    for topic_idx, topic in enumerate(nmf.components_):
        print 'Topic #{}:' .format(topic_idx)
        print ' '.join([tfidf_vectorizer.get_feature_names()[i] for i \
                            in topic.argsort()[:-n_topic_words -1:-1]])
        print()
    return str(n_topics) + 'topics' 

def ldatopics(n_topics):
    corpus = []
    stemmer = SnowballStemmer('english')
    for prodId in df.productId.unique():
        docs = []
        for doc in df[df.productId == prodId].text:
            if doc:
                doc = ' '.join(stemmer.stem(word.strip(punctuation).lower()) \
                                                for word in doc.split())
                docs.append(doc)
        corpus.append(' '.join(docs))
    tfidf_vectorizer = TfidfVectorizer(stop_words='english')
    tfidf = tfidf_vectorizer.fit_transform(corpus)
    lda = LatentDirichletAllocation(n_topics=n_topics, max_iter=5,
                                learning_method='online',
                                learning_offset=50.,
                                random_state=0).fit(tfidf)
    n_topic_words = n_topics
    for topic_idx, topic in enumerate(lda.components_):
        print 'Topic #{}:' .format(topic_idx)
        print ' '.join([tfidf_vectorizer.get_feature_names()[i] for i \
                            in topic.argsort()[:-n_topic_words -1:-1]])
        print()
    return str(n_topics) + 'topics' 


In [55]:
type(tfidf)

NameError: name 'tfidf' is not defined

In [53]:
ldatopics(13)

Topic #0:
product workout pre tast pump great good feel energi use tri ani work
()
Topic #1:
protein product flavor bi great tast caffein differ mix use good redic watermelon
()
Topic #2:
amino tast great tri good cheap product flavor pretti protein trutein digest add
()
Topic #3:
product great workout tast energi protein tri flavor use dure like good mix
()
Topic #4:
phone lean protein monstrous mix esp galaxi nevvvvver idgaf relato gainer somwth tast
()
Topic #5:
bar product like tast great tri ape jerki protein mom good chewi small
()
Topic #6:
fish protein product oil great burp like result tast time junk good bcaa
()
Topic #7:
tast protein mix great product like good use flavor fish tri workout love
()
Topic #8:
product did great return work multi good result noth conveni mix use tast
()
Topic #9:
brazil recomendo bom melhor muito great product horribl affect primarili use stomach indian
()
Topic #10:
pre great effect excel pill workout product focus size defen break met swallow
(

'13topics'

In [58]:
topics(12)

Topic #0:
gain product notic week use muscl day test strength result mass increas
()
Topic #1:
protein chocol mix milk whey powder vanilla tast shake great flavor best
()
Topic #2:
workout pre pump energi focus scoop feel tri preworkout caffein use great
()
Topic #3:
bar protein tast chocol quest eat like snack good textur cooki flavor
()
Topic #4:
vitamin multi multivitamin pill day great price need feel swallow easi energi
()
Topic #5:
sleep night wake zma asleep help feel dream fall rest bed befor
()
Topic #6:
bcaa workout recoveri amino glutamin product great post use help sore dure
()
Topic #7:
tast flavor like mix good water tri drink bad just scoop pretti
()
Topic #8:
fish oil burp omega fishi dha epa pill cla great softgel high
()
Topic #9:
creatin mix unflavor strength monohydr gain water shake price use mono notic
()
Topic #10:
product good great veri price work excel qualiti recommend use cheap best
()
Topic #11:
fat energi burner cla help work weight day loss appetit feel p

'12topics'

gainer, protein, preworkout, snack, vitamin, sleep, bcaa, mixturability, fish oil, creatine, value, fat burner

In [196]:
topics(13)

Topic #0:
product notic use day week work differ test feel ani strength booster gain
()
Topic #1:
protein chocol mix milk whey powder vanilla tast shake flavor great best use
()
Topic #2:
workout pre pump energi focus scoop feel preworkout tri caffein great use c4
()
Topic #3:
bar protein tast chocol quest eat like snack good textur cooki flavor chewi
()
Topic #4:
vitamin multi multivitamin pill day great price need swallow easi energi feel pak
()
Topic #5:
sleep night wake zma asleep help dream feel fall rest bed hour melatonin
()
Topic #6:
bcaa workout recoveri amino glutamin product great post use help sore dure mix
()
Topic #7:
tast flavor like mix good water tri drink bad just scoop realli pretti
()
Topic #8:
fish oil burp omega fishi dha epa pill cla great softgel high supplement
()
Topic #9:
creatin mix unflavor strength monohydr gain water shake price use notic muscl mono
()
Topic #10:
product good great veri price work excel qualiti recommend use cheap best high
()
Topic #11:


'13topics'

strength gaib, fat burner, value, creatine, fish oil, quality/mixturability, bcaa, sleep, vitamin, bar, preworkout, protein, mass gainer

In [56]:
topics(14)

Topic #0:
product notic day week use differ test work feel booster strength ani gain increas
()
Topic #1:
protein chocol mix milk whey powder tast vanilla flavor shake great best tri use
()
Topic #2:
workout pre pump energi focus scoop feel preworkout tri caffein use c4 like gym
()
Topic #3:
bar protein tast chocol quest eat like snack textur cooki flavor chewi great tri
()
Topic #4:
vitamin multi multivitamin pill day need swallow great easi feel price pak energi best
()
Topic #5:
sleep night wake zma asleep help dream feel fall rest bed hour melatonin deep
()
Topic #6:
bcaa workout recoveri amino glutamin post sore help dure mix use muscl day drink
()
Topic #7:
tast flavor like mix water drink tri just bad scoop sweet don realli green
()
Topic #8:
fish oil burp omega fishi dha epa pill cla softgel high tast supplement qualiti
()
Topic #9:
creatin mix unflavor strength monohydr gain water shake price notic muscl mono use creapur
()
Topic #10:
product great work recommend excel use bes

'14topics'

In [ ]:
value, gainer, fat burner, value,  creatine, fish oil, mixturability, bcaa, sleep, vitamin, bar, preworkout, protein, strength

In [395]:
df[df.productId == 'prod2720004']

brandName     brandId        name    productId         username  \
14489  Miracle Noodle  cat1840001  Angel Hair  prod2720004       spacejim67   
14490  Miracle Noodle  cat1840001  Angel Hair  prod2720004  kimberlygorgone   
14491  Miracle Noodle  cat1840001  Angel Hair  prod2720004       Brit0Brit0   

          userId                       _id  height  weight  bodyfat  \
14489   57009792  573d267c0cf223211f76869a    70.0   191.0     12.8   
14490  123245611  568eef0d0cf2b1f59d30c8c8    69.0   188.0     25.5   
14491  104394142  55d869220cf2153e8e940538    71.0   198.0     11.0   

      totalItems                                               text  \
14489       None  So I gave this a ten because, it had no bad sm...   
14490       None  Do not buy this!!!! It's like chewing a rubber...   
14491       None  One of the best things I could have purchased....   

                                     date                             modDate  \
14489  {u'$numberLong': u'1463625340000'}  {u'$numberLong': u'1463625340000'}   
14490  {u'$numberLong': u'1452207885000'}  {u'$numberLong': u'1452207885000'}   
14491  {u'$numberLong': u'1440246050000'}  {u'$numberLong': u'1440246050000'}   

                  slug  verifiedBuyerRating  \
14489       spacejim67                  7.0   
14490  kimberlygorgone                  7.0   
14491       Brit0Brit0                  7.0   

                                             description updateStatusReason  \
14489  Low Calorie, Soy Free, Gluten Free, Vegan, Non...               None   
14490  Low Calorie, Soy Free, Gluten Free, Vegan, Non...               None   
14491  Low Calorie, Soy Free, Gluten Free, Vegan, Non...               None   

       rating  
14489    10.0  
14490     1.0  
14491    10.0

In [3]:
df

brandName         brandId                  name    productId  \
0        360CUT      cat1150004               360LEAN  prod1620022   
1        360CUT      cat1150004               360LEAN  prod1620022   
2        360CUT      cat1150004               360LEAN  prod1620022   
3        360CUT      cat1150004               360LEAN  prod1620022   
4        360CUT      cat1150004               360LEAN  prod1620022   
5        360CUT      cat1150004               360LEAN  prod1620022   
6        360CUT      cat1150004               360LEAN  prod1620022   
7        360CUT      cat1150004               360LEAN  prod1620022   
8        360CUT      cat1150004               360LEAN  prod1620022   
9        360CUT      cat1150004               360LEAN  prod1620022   
10       360CUT      cat1150004               360LEAN  prod1620022   
11       360CUT      cat1150004               360LEAN  prod1620022   
12       360CUT      cat1150004               360LEAN  prod1620022   
13       360CUT      cat1150004               360LEAN  prod1620022   
14       360CUT      cat1150004               360LEAN  prod1620022   
15       360CUT      cat1150004               360LEAN  prod1620022   
16       360CUT      cat1150004               360LEAN  prod1620022   
17       360CUT      cat1150004               360LEAN  prod1620022   
18       360CUT      cat1150004                360PRE  prod1620026   
19       360CUT      cat1150004                360PRE  prod1620026   
20       360CUT      cat1150004                360PRE  prod1620026   
21       360CUT      cat1150004                360PRE  prod1620026   
22       360CUT      cat1150004                360PRE  prod1620026   
23       360CUT      cat1150004                360PRE  prod1620026   
24       360CUT      cat1150004                360PRE  prod1620026   
25       360CUT      cat1150004                360PRE  prod1620026   
26       360CUT      cat1150004                360PRE  prod1620026   
27       360CUT      cat1150004                360PRE  prod1620026   
28       360CUT      cat1150004                360PRE  prod1620026   
29       360CUT      cat1150004                360PRE  prod1620026   
...         ...             ...                   ...          ...   
67581  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67582  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67583  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67584  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67585  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67586  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67587  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67588  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67589  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67590  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67591  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67592  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67593  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67594  Dymatize  BRAND_DYMATIZE     Super Mass Gainer        23745   
67595    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67596    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67597    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67598    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67599    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67600    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67601    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67602    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67603    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67604    S.A.N.    BRAND_S_A_N_  Green Supreme Fusion   prod680032   
67605       NOW       BRAND_NOW         Ginkgo Biloba        23649   
67606       NOW 

In [16]:
df2 = pd.read_pickle('nutrition.pkl')
sf = gl.SFrame(df2[['userId', 'productId', 'rating']])
user_info = gl.SFrame(df2[['userId', 'height', 'weight', 'bodyfat']])
item_info = gl.SFrame(df2[['brandId', 'productId', 'description']])
user_info = user_info.fillna('weight',0)
user_info = user_info.fillna('height',0)
user_info = user_info.fillna('bodyfat',0)
m = factorization_recommender.create(sf, user_id='userId', \
                                    item_id='productId', target='rating',\
                                user_data=user_info, item_data=item_info)

Recsys training: model = factorization_recommender

Preparing data set.

Data has 64312 observations with 39465 users and 2131 items.

Data prepared in: 0.268693s

Training factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 8        |

| regularization                 | L2 Regularization on Factors                     | 1e-08    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-10    |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 50       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 10000 / 64312 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 7.14286           | Not Viable                               |

| 1       | 1.78571           | Not Viable                               |

| 2       | 0.446429          | Not Viable                               |

| 3       | 0.111607          | 2.4037                                   |

| 4       | 0.0558036         | 2.85335                                  |

| 5       | 0.0279018         | 3.61804                                  |

| 6       | 0.0139509         | 4.141                                    |

+---------+-------------------+------------------------------------------+

| Final   | 0.111607          | 2.4037                                   |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training RMSE | Step Size   |

+---------+--------------+-------------------+-----------------------+-------------+

| Initial | 105us        | 4.54328           | 2.1315                |             |

+---------+--------------+-------------------+-----------------------+-------------+

| 1       | 142.861ms    | 5.05691           | 2.24874               | 0.111607    |

| 2       | 249.084ms    | 4.1686            | 2.0417                | 0.111607    |

| 3       | 353.507ms    | 3.32151           | 1.82248               | 0.111607    |

| 4       | 433.281ms    | 2.81503           | 1.67778               | 0.111607    |

| 5       | 560.578ms    | 2.40009           | 1.54919               | 0.111607    |

| 6       | 811.301ms    | 2.08135           | 1.44265               | 0.111607    |

| 10      | 1.21s        | 1.40615           | 1.18576               | 0.111607    |

| 11      | 1.33s        | 1.2831            | 1.13269               | 0.111607    |

| 20      | 2.22s        | 0.893228          | 0.945035              | 0.111607    |

| 30      | 3.02s        | 0.682877          | 0.826276              | 0.111607    |

| 40      | 3.80s        | 0.961171          | 0.980317              | 0.111607    |

| 50      | 4.61s        | 0.513387          | 0.716403              | 0.111607    |

+---------+--------------+-------------------+-----------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training RMSE.

Final objective value: 0.719539

Final training RMSE: 0.848166

In [54]:
user = random.randint(1,1000000000)
recent_data = gl.SFrame({'userId':[1, 1, 1], \
                      'productId':['prod2810009', 'prod3280025', 'prod1620046'], \
                         'rating':[8, 7, 2], \
                       'username':['test', 'test', 'test'], \
                         'height': [60, 60, 60], \
                         'weight': [165, 165, 165], \
                        'bodyfat': [10, 10, 10]})

recs = m.recommend(users=[1], new_observation_data=recent_data)

In [55]:
recs

Columns:
	userId	int
	productId	str
	score	float
	rank	int

Rows: 10

Data:
+--------+-------------+---------------+------+
| userId |  productId  |     score     | rank |
+--------+-------------+---------------+------+
|   1    |    23899    | 10.9215935248 |  1   |
|   1    |    28105    | 10.8611961084 |  2   |
|   1    | prod4030010 | 10.8009966391 |  3   |
|   1    | prod1700096 | 10.6903678107 |  4   |
|   1    | prod1330052 | 10.2422781366 |  5   |
|   1    | prod1010022 | 10.1725648421 |  6   |
|   1    | prod2030106 | 10.1561737883 |  7   |
|   1    | prod3200011 | 10.1505964774 |  8   |
|   1    | prod2320060 | 10.1493489819 |  9   |
|   1    |    26102    | 10.1187886375 |  10  |
+--------+-------------+---------------+------+
[10 rows x 4 columns]

In [38]:
recs['productId']

dtype: str
Rows: 10
['23899', '28105', 'prod4030010', 'prod1620046', 'prod1700096', 'prod1330052', 'prod1010022', 'prod2030106', 'prod3200011', 'prod2320060']

In [42]:
df = pd.read_pickle('rateaproduct.pkl')

Index([u'productId', u'drop_down'], dtype='object')

In [51]:
top = []
for prod in recs['productId']:
    top.append(df[df.productId == prod].values[0][1])

In [53]:
for rec in top:
    print rec

High Energy Labs Multiflex Complete
Eggology 100% Egg Whites
Neogenix Bodyforge Ultra
RockTape Kinesiology Tape
Inner Armour Shred Factor
iForce Nutrition Intimidate SRT
BarnDad Innovative Nutrition Fiber DX
Probar Core Protein Bar
Core Nutritionals Pump
Axis Labs Instantized N'Gage Amino


In [31]:
df[df.productId =='prod4030010']

brandName         brandId             name    productId     username  \
46143  Neogenix  BRAND_NEOGENIX  Bodyforge Ultra  prod4030010  Dairyman209   
46144  Neogenix  BRAND_NEOGENIX  Bodyforge Ultra  prod4030010   alexaluker   

          userId                       _id  height  weight  bodyfat  \
46143   52412341  5814a0500cf2dadf1bd4921e     NaN   160.0     14.5   
46144  140223911  581234f50cf236d07b65ea7a    70.0   177.0     15.0   

      totalItems                                               text  \
46143       None                                                      
46144       None  Flavor is atrocious. Only when I mixed it with...   

                                     date                             modDate  \
46143  {u'$numberLong': u'1477746768000'}  {u'$numberLong': u'1477746768000'}   
46144  {u'$numberLong': u'1477588213000'}  {u'$numberLong': u'1477588213000'}   

              slug  verifiedBuyerRating  \
46143  Dairyman209                  9.0   
46144   alexaluker                  9.0   

                                             description updateStatusReason  \
46143  Designed to Enhance Strength, Endurance and Re...               None   
46144  Designed to Enhance Strength, Endurance and Re...               None   

       rating  
46143    10.0  
46144     8.0

In [160]:
df.to_pickle('nutrition.pkl')

In [42]:
df2 =df2[df2.productId != 'prod540102']

In [43]:
df2.to_pickle('rateaproduct.pkl')

In [44]:
df2.shape

(2130, 2)

In [143]:
df.shape

(61550, 19)

In [158]:
df =df[df.name != 'Xercuffs']

In [159]:
df.shape

(61524, 19)